community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/CheckPointerImpl.java

```diff
@@ -0,0 +1,158 @@
+        this.msgLog = logProvider.getLog( CheckPointerImpl.class );
+        this.tracer = tracer;
+    }
+
+    @Override
+    public void start() throws Throwable
+    {
+        this.channel = logFile.getWriter();
+        this.transactionLogWriter =
+                new TransactionLogWriter( new LogEntryWriterV1( channel, NeoCommandHandler.EMPTY ) );
+    }
+
+    @Override
+    public void forceCheckPoint() throws IOException
+    {
+        synchronized ( logFile )
+        {
+            doCheckPoint();
+        }
+    }
+
+    @Override
+    public void checkPointIfNeeded()
+    {
+        if ( threshold.isCheckPointingNeeded() )
+        {
+            synchronized ( logFile )
+            {
+                if ( threshold.isCheckPointingNeeded() )
+                {
+                    try ( LogCheckPointEvent event = tracer.beginCheckPoint() )
+                    {
+                        try
+                        {
+                            doCheckPoint();
+                        }
+                        catch ( IOException ex )
+                        {
+                            throw new UnderlyingStorageException( ex );
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private void doCheckPoint() throws IOException
+    {
+        long[] lastCommittedTransaction = transactionIdStore.getLastCommittedTransaction();
+        long lastCommittedTransactionId = lastCommittedTransaction[0];
+        LogPosition logPosition =
+                new LogPosition( lastCommittedTransaction[2], lastCommittedTransaction[3] );
+        msgLog.info( prefix( lastCommittedTransactionId ) + "Starting check pointing..." );
+
+        /*
+         * Check kernel health before going into waiting for transactions to be closed, to avoid
+         * getting into a scenario where we would await a condition that would potentially never
+         * happen.
+         */
+        kernelHealth.assertHealthy( IOException.class );
+
+        /*
+         * First we flush the store. If we fail now or during the flush, on recovery we'll find the
+         * earlier check point and replay from there all the log entries. Everything will be ok.
+         */
+        msgLog.info( prefix( lastCommittedTransactionId ) + " Starting store flush..." );
+        storeFlusher.forceEverything();
+
+        synchronized ( channel ) // TODO: is this needed?
+        {
+            /*
+             * Check kernel health before going to write the next check point.
+             * In case of a panic this rotation will be aborted, which is the safest alternative
+             * so that the next recovery will have a chance to repair the damages.
+             */
+            kernelHealth.assertHealthy( IOException.class );
+            transactionLogWriter.checkPoint( logPosition );
+            channel.emptyBufferIntoChannelAndClearIt();
+            channel.force();
+        }
```

Comment:
I wonder if you could get the BatchingTransactionAppender to write the check-point entry for you, so you go through the write-path where concurrent `channel.force()` calls gets coalesced.